File: concepts/Functional_data_structure/Stack/Stack.java
public class Stack {
  private int element;
  private Stack next;
  public static final Stack empty = null;
  public static Stack push(int x,Stack s) {
    return new Stack(x,s);
  }
  public static int top(Stack s) {
    return s.element;
  }
  public static Stack pop(Stack s) {
    return s.next;
  }
  private Stack(int element, Stack next) {
    this.element = element;
    this.next = next;
  }
}

File: concepts/Functional_data_structure/Stack/Haskell/Stack.hs
-- Source: Chapter 40: Functional Data Structures by C. Okasaki. In: Handbook of Data Structures and Applications. Chapman & Hall/CRC.

data Stack = Empty | Push Int Stack 

empty = Empty
push x s  = Push x s
top (Push x s) = x
pop (Push x s) = s
